test/test_runlib.py

```diff
@@ -22,24 +22,126 @@
+    expected = ["foo"]
+    result = _apply_exclude_patterns(names, patterns)
+    self.assertListEqual(sorted(result), sorted(expected))
+
+
+class TestRecordArtifactsAsDict(unittest.TestCase):
+  """Test record_artifacts_as_dict(artifacts). """
+
+  @classmethod
+  def setUpClass(self):
+    """Create and change into temp test directory with dummy artifacts. """
+    # Clear user set excludes
+    settings.ARTIFACT_EXCLUDES = []
+
+    self.test_dir = tempfile.mkdtemp()
+    os.chdir(self.test_dir)
+    open("foo", "w").write("foo")
+    open("bar", "w").write("bar")
+
+    os.mkdir("subdir")
+    os.mkdir("subdir/subsubdir")
+    open("subdir/foosub", "w").write("foosub")
+    open("subdir/subsubdir/foosubsub", "w").write("foosubsub")
+
+  @classmethod
+  def tearDownClass(self):
+    """Change back to initial working dir and remove temp test directory. """
+    os.chdir(WORKING_DIR)
+    shutil.rmtree(self.test_dir)
+
+  def test_empty_artifacts_list_record_nothing(self):
+    """Empty list passed. Return empty dict. """
+    self.assertDictEqual(record_artifacts_as_dict([]), {})
+
+  def test_not_existing_artifacts_in_list_record_nothing(self):
+    """List with not existing artifact passed. Return empty dict. """
+    self.assertDictEqual(record_artifacts_as_dict(["baz"]), {})
+
+  def test_record_dot_check_files_hash_dict_schema(self):
+    """Traverse dir and subdirs. Record three files. """
+    artifacts_dict = record_artifacts_as_dict(["."])
+
+    for key, val in artifacts_dict.iteritems():
+      ssl_crypto.formats.HASHDICT_SCHEMA.check_match(val)
+
+    self.assertListEqual(sorted(artifacts_dict.keys()),
+      sorted(["foo", "bar", "subdir/foosub", "subdir/subsubdir/foosubsub"]))
+
+  def test_record_files_and_subdirs(self):
+    """Explicitly record files and subdirs. """
+    artifacts_dict = record_artifacts_as_dict(["foo", "subdir"])
+
+    for key, val in artifacts_dict.iteritems():
+      ssl_crypto.formats.HASHDICT_SCHEMA.check_match(val)
+
+    self.assertListEqual(sorted(artifacts_dict.keys()),
+      sorted(["foo", "subdir/foosub", "subdir/subsubdir/foosubsub"]))
+
+  def test_record_dot_exclude_foo_star_from_recording(self):
+    """Traverse dir and subdirs from. Exclude pattern. Record one file. """
+    settings.ARTIFACT_EXCLUDES = ["foo*"]
+    artifacts_dict = record_artifacts_as_dict(["."])
+
+    ssl_crypto.formats.HASHDICT_SCHEMA.check_match(artifacts_dict["bar"])
+    self.assertListEqual(artifacts_dict.keys(), ["bar"])
+    settings.ARTIFACT_EXCLUDES = []
+
+  def test_exclude_subsubdir(self):
+    """Traverse dir and subdirs. Exclude subsubdir. """
+    settings.ARTIFACT_EXCLUDES = ["subsubdir"]
+    artifacts_dict = record_artifacts_as_dict(["."])
+
+    for key, val in artifacts_dict.iteritems():
+      ssl_crypto.formats.HASHDICT_SCHEMA.check_match(val)
+
+    self.assertListEqual(sorted(artifacts_dict.keys()),
+        sorted(["foo", "bar", "subdir/foosub"]))
+    settings.ARTIFACT_EXCLUDES = []
```

Comment:
I wonder if it would be a good idea to have this under a `try:finally:` or a `TearDown()` so that clearing the excludes is always performed and doesn't pollute ensuing tests.